ESP8266/ESP8266.cpp

```diff
@@ -16,14 +16,22 @@
 
 #include "ESP8266.h"
 
+#define ATPARSER_BUFSIZE    256  //bytes
+#define ATPARSER_TIMEOUT    8000 //ms
+#define BAUDRATE            115200
+
 ESP8266::ESP8266(PinName tx, PinName rx, bool debug)
-    : _serial(tx, rx, 1024), _parser(_serial)
-    , _packets(0), _packets_end(&_packets)
+    :  _packets(0), _packets_end(&_packets)
 {
-    _serial.baud(115200);
-    _parser.debugOn(debug);
+    _fh = new BufferedSerial(tx, rx, BAUDRATE);
+    _parser =  new ATParser(*_fh, ATPARSER_BUFSIZE, ATPARSER_TIMEOUT, debug);
```

Comment:
I realized ATParser is taking a reference, I moved that issue over to [here](https://github.com/ARMmbed/mbed-os/pull/4119/files#r113273995).

As for the dynamic memory allocation, I'm still confused on what that is accomplishing. Do you need to clean up the memory outside of the life of the enclosing class? As it is the dynamic memory will add a layer of indirection and just make it harder to statical determine your module's memory consumption.